File: src/main/java/net/jgp/books/spark/ch10/lab930_memory_sink/StreamRecordInMemoryApp.java
package net.jgp.books.spark.ch10.lab930_memory_sink;

import java.util.concurrent.TimeoutException;

import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.streaming.OutputMode;
import org.apache.spark.sql.streaming.StreamingQuery;
import org.apache.spark.sql.types.StructType;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import net.jgp.books.spark.ch10.x.utils.streaming.lib.StreamingUtils;

/**
 * Analyzes the records on the stream and send each record to a debugger
 * class.
 * 
 * @author jgp
 *
 */
public class StreamRecordInMemoryApp {
  private static Logger log =
      LoggerFactory.getLogger(StreamRecordInMemoryApp.class);

  public static void main(String[] args) {
    StreamRecordInMemoryApp app = new StreamRecordInMemoryApp();
    try {
      app.start();
    } catch (TimeoutException e) {
      log.error("A timeout exception has occured: {}", e.getMessage());
    }
  }

  private void start() throws TimeoutException {
    log.debug("-> start()");

    SparkSession spark = SparkSession.builder()
        .appName("Read lines over a file stream")
        .master("local")
        .getOrCreate();

    StructType recordSchema = new StructType()
        .add("fname", "string")
        .add("mname", "string")
        .add("lname", "string")
        .add("age", "integer")
        .add("ssn", "string");

    Dataset<Row> df = spark
        .readStream()
        .format("csv")
        .schema(recordSchema)
        .csv(StreamingUtils.getInputDirectory());

    StreamingQuery query = df
        .writeStream()
        .outputMode(OutputMode.Append())
        .format("memory") // #A
        .option("queryName", "people") // #B
        .start();

    // Wait and process the incoming stream for the next minute
    Dataset<Row> queryInMemoryDf; // #C
    int iterationCount = 0;
    long start = System.currentTimeMillis();
    while (query.isActive()) { // #D
      queryInMemoryDf = spark.sql("SELECT * FROM people"); // #E
      iterationCount++;
      log.debug("Pass #{}, dataframe contains {} records",
          iterationCount,
          queryInMemoryDf.count());
      queryInMemoryDf.show();
      if (start + 60000 < System.currentTimeMillis()) {
        query.stop(); // #F
      }
      try {
        Thread.sleep(2000);
      } catch (InterruptedException e) {
        // Simply ignored
      }
    }

    log.debug("<- start()");
  }
}

File: src/main/python/lab930_memory_sink/streamRecordInMemoryApp.py
"""
  Analyzes the records on the stream and send each record to a debugger class.

  @author rambabu.posa
"""
from pyspark.sql import SparkSession
from pyspark.sql.types import StructType, StructField, StringType, IntegerType
import logging
import time

logging.debug("-> start")

# Creates a session on a local master
spark = SparkSession.builder.appName("Read lines over a file stream") \
    .master("local[*]").getOrCreate()

# The record structure must match the structure of your generated
# record
# (or your real record if you are not using generated records)
recordSchema = StructType([StructField('fname', StringType(), True),
                           StructField('mname', StringType(), True),
                           StructField('lname', StringType(), True),
                           StructField('age', IntegerType(), True),
                           StructField('ssn', StringType(), True)])

# Reading the record is always the same
df = spark.readStream.format("csv") \
    .schema(recordSchema) \
    .csv("/tmp/")
# use for windows
#   .csv("C:/tmp/")

query = df.writeStream.outputMode("append") \
    .format("memory") \
    .option("queryName", "people") \
    .start()

iterationCount = 0
start = int(round(time.time() * 1000))

while query.isActive:
    queryInMemoryDf = spark.sql("SELECT * FROM people")
    iterationCount += 1
    logging.debug("Pass #{}, dataframe contains {} records".format(iterationCount, queryInMemoryDf.count()))

    queryInMemoryDf.show()

    end = int(round(time.time() * 1000))

    if start + 60000 < end:
        query.stop()

    time.sleep(2)

logging.debug("<- end")
